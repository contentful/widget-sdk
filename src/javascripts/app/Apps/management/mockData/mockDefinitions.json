[
  {
    "sys": {
      "id": "3AjEyjWz5tRouW4cVOF9la",
      "type": "AppDefinition",
      "organization": {
        "sys": {
          "type": "Link",
          "linkType": "Organization",
          "id": "2b5UqceLpT58d9tat2os42"
        }
      },
      "createdAt": "2020-01-17T12:15:36.351Z",
      "updatedAt": "2020-01-17T12:27:53.592Z",
      "createdBy": {
        "sys": {
          "type": "Link",
          "linkType": "User",
          "id": "5NY5T2MqqnqTlRlJ022S9m"
        }
      },
      "updatedBy": {
        "sys": {
          "type": "Link",
          "linkType": "User",
          "id": "5NY5T2MqqnqTlRlJ022S9m"
        }
      }
    },
    "name": "Google Analytics",
    "src": "http://localhost:1234",
    "locations": [
      {
        "location": "app-config"
      },
      {
        "location": "entry-sidebar"
      }
    ]
  },
  {
    "sys": {
      "id": "2WoMzSs2fRJx79JeVvqdGR",
      "type": "AppDefinition",
      "organization": {
        "sys": {
          "type": "Link",
          "linkType": "Organization",
          "id": "2b5UqceLpT58d9tat2os42"
        }
      },
      "createdAt": "2019-11-24T15:11:07.998Z",
      "updatedAt": "2019-11-24T15:11:07.998Z",
      "createdBy": {
        "sys": {
          "type": "Link",
          "linkType": "User",
          "id": "2jvc3kU4n7OIABiFMTaGyB"
        }
      },
      "updatedBy": {
        "sys": {
          "type": "Link",
          "linkType": "User",
          "id": "2jvc3kU4n7OIABiFMTaGyB"
        }
      }
    },
    "name": "Jira",
    "src": "http://localhost:1234",
    "locations": [
      {
        "location": "app-config"
      },
      {
        "location": "entry-sidebar"
      }
    ]
  },
  {
    "sys": {
      "id": "3cG9ueyzkY0GCndLiQzmZ0",
      "type": "AppDefinition",
      "organization": {
        "sys": {
          "type": "Link",
          "linkType": "Organization",
          "id": "2b5UqceLpT58d9tat2os42"
        }
      },
      "createdAt": "2020-01-17T13:40:14.180Z",
      "updatedAt": "2020-01-28T09:44:03.251Z",
      "createdBy": {
        "sys": {
          "type": "Link",
          "linkType": "User",
          "id": "2jvc3kU4n7OIABiFMTaGyB"
        }
      },
      "updatedBy": {
        "sys": {
          "type": "Link",
          "linkType": "User",
          "id": "2jvc3kU4n7OIABiFMTaGyB"
        }
      }
    },
    "name": "Smartling",
    "src": "http://localhost:8080/frontend",
    "locations": [
      {
        "location": "app-config"
      },
      {
        "location": "entry-sidebar"
      }
    ]
  }
]
