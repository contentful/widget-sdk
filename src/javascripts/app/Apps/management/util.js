import { getSpaces } from 'services/TokenStore';
import { createSpaceEndpoint } from 'data/EndpointFactory';
import { create as createSpaceEnvRepo } from 'data/CMA/SpaceEnvironmentsRepo';
import * as urlSafeBase64 from 'urlsafe-base64';
import moment from 'moment';

export async function getOrgSpacesFor(orgId) {
  const spaces = await getSpaces();

  return spaces.filter((s) => s.organization.sys.id === orgId);
}

export async function getEnvsFor(spaceId) {
  const spaceEndpoint = createSpaceEndpoint(spaceId, 'master');
  const spaceEnvRepo = createSpaceEnvRepo(spaceEndpoint);
  const { environments } = await spaceEnvRepo.getAll();

  return environments;
}

export function keyPemTobase64Der(pem) {
  const base64Der = pem.replace(/-*(BEGIN|END) PUBLIC KEY-*/g, '').replace(/\s+/g, '');

  return base64Der;
}

export async function getSha256FromBase64(base64String) {
  const binaryDer = window.atob(base64String);
  const derArray = new Uint8Array(binaryDer.length);

  for (let i = 0; i < binaryDer.length; i++) {
    derArray[i] = binaryDer.charCodeAt(i);
  }

  const fingerprintArray = new Uint8Array(await window.crypto.subtle.digest('SHA-256', derArray));
  let fingerprintBinary = '';

  for (let i = 0; i < fingerprintArray.length; i++) {
    fingerprintBinary += String.fromCharCode(fingerprintArray[i]);
  }

  return fingerprintBinary;
}

export function base64ToHex(base64String) {
  const u8Array = urlSafeBase64.decode(base64String);

  let hex = '';

  for (let i = 0; i < u8Array.length; i++) {
    hex += u8Array[i].toString(16).padStart(2, '0');
  }

  return hex;
}

export function formatPastDate(date) {
  const today = moment().startOf('day');
  const pastDay = moment(date).startOf('day');
  const dayDiff = today.diff(pastDay, 'days');

  if (dayDiff == 0) {
    return 'today';
  }

  if (dayDiff === 1) {
    return 'yesterday';
  }

  const thisWeek = moment().startOf('week');
  const pastWeek = moment(date).startOf('week');
  const weekDiff = thisWeek.diff(pastWeek, 'weeks');

  if (dayDiff < 7 || weekDiff === 0) {
    return dayDiff + ' days ago';
  }

  if (weekDiff == 1) {
    return 'last week';
  }

  const thisMonth = moment().startOf('month');
  const pastMonth = moment(date).startOf('month');
  const monthDiff = thisMonth.diff(pastMonth, 'months');

  if (weekDiff < 4 || monthDiff === 0) {
    return weekDiff + ' weeks ago';
  }

  if (monthDiff === 1) {
    return 'last month';
  }

  return monthDiff + 'months ago';
}
