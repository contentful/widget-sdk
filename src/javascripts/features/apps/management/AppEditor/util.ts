import { isEqual } from 'lodash';

export function arrayStartsWith<T>(array: T[], target: T[]): boolean {
  return isEqual(array.slice(0, target.length), target);
}
