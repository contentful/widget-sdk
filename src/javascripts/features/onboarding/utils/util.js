import { getSpaceContext } from 'classes/spaceContext';
import * as TokenStore from 'services/TokenStore';
import { getCMAClient } from 'core/services/usePlainCMAClient';
import { captureWarning } from 'core/monitoring';
import { getBrowserStorage } from 'core/services/BrowserStorage';
import { getUser } from 'services/TokenStore';
import { router } from 'core/react-routing';
import { showReplaceSpaceWarning } from '../components/ReplaceSpaceDialog';
import {
  markSpace,
  getStoragePrefix as getOnboardingStoragePrefix,
  MODERN_STACK_ONBOARDING_SPACE_NAME,
  DEFAULT_LOCALE,
} from 'components/shared/auto_create_new_space/CreateModernOnboardingUtils';
import { getQueryString } from 'utils/location';
import { fetchUserData } from 'features/user-profile';
import { tracking, defaultEventProps } from 'analytics/Analytics';
import { getVariation, FLAGS } from 'core/feature-flags';

const store = getBrowserStorage();

const getDevStoragePrefix = async () => {
  const user = await getUser();
  return `ctfl:${user.sys.id}:persona`;
};

const getStoragePrefix = async () => {
  const user = await getUser();
  return `ctfl:${user.sys.id}:exploreOnboardingSeen`;
};

export const markExploreOnboardingSeen = async () => {
  const prefix = await getStoragePrefix();
  store.set(prefix, true);
};

export const hasSeenExploreOnboarding = async () => {
  const prefix = await getStoragePrefix();
  return store.get(prefix);
};

export const CONTROL_EXP_VARIATION = 'control';
export const TREATMENT_EXP_VARIATION = 'flexible-onboarding';
export const PREASSIGN_ONBOARDING_EXP_VARIATION = 'preassign-onboarding';
export const BLANK_SPACE_NAME = 'Blank';

export const renameSpace = async (newName, spaceId) => {
  const spaceContext = getSpaceContext();
  const currentSpace = await TokenStore.getSpace(spaceId);
  const updatedSpace = { sys: currentSpace.sys, name: newName };
  const cma = getCMAClient({ spaceId });

  const asyncError = new Error('Something went wrong while updating space');
  try {
    await cma.space.update({}, updatedSpace);
  } catch (err) {
    captureWarning(asyncError, {
      extra: {
        message: err.message,
      },
    });
  }

  await TokenStore.refresh();
  const newSpace = await TokenStore.getSpace(spaceId);
  await spaceContext.resetWithSpace(newSpace);

  return newSpace;
};

export const handleReplaceSpace = (currentSpaceId) => {
  showReplaceSpaceWarning(currentSpaceId, (spaceId) => {
    markSpace(spaceId);
    renameSpace(MODERN_STACK_ONBOARDING_SPACE_NAME, spaceId);
    router.navigate({
      path: 'spaces.detail.onboarding.copy',
      spaceId,
    });
  });
};

export const handleGetStarted = (currentSpaceId) => {
  markExploreOnboardingSeen();
  renameSpace(MODERN_STACK_ONBOARDING_SPACE_NAME, currentSpaceId);
};

export const isDeveloper = async () => {
  const prefix = await getDevStoragePrefix();
  const storageDevValue = store.get(prefix);

  if (storageDevValue) {
    return storageDevValue ? true : false;
  } else {
    const user = await fetchUserData();
    const hasGithubIdentityConnected =
      user?.identities?.filter((i) => i.provider === 'github').length >= 1;

    const { persona } = getQueryString();

    const devValue = persona === 'developer' || hasGithubIdentityConnected;

    if (devValue) {
      store.set(prefix, 'developer');
    }

    return devValue;
  }
};

export const goToDeveloperOnboarding = async ({ org, markOnboarding }) => {
  const orgId = org.sys.id;
  const client = getCMAClient();
  const newSpace = await client.space.create(
    {
      organizationId: orgId,
    },
    {
      name: MODERN_STACK_ONBOARDING_SPACE_NAME,
      defaultLocale: DEFAULT_LOCALE,
    }
  );

  markOnboarding();
  markSpace(newSpace.sys.id);

  store.set(`${getOnboardingStoragePrefix()}:currentStep`, {
    path: 'spaces.detail.onboarding.getStarted',
    params: {
      spaceId: newSpace.sys.id,
    },
  });

  await TokenStore.refresh();

  router.navigate({
    path: 'spaces.detail.onboarding.getStarted',
    spaceId: newSpace.sys.id,
  });
};

export const trackContentTypeSave = async ({ organizationId }) => {
  const enabledPreassignExperiment = await getVariation(
    FLAGS.EXPERIMENT_PREASSIGN_ONBOARDING_FLOW,
    {
      organizationId,
    }
  );
  const isDev = await isDeveloper();
  if (enabledPreassignExperiment !== null) {
    if (isDev) {
      tracking.experimentGoalAchieved({
        ...defaultEventProps(),
        experiment_id: FLAGS.EXPERIMENT_PREASSIGN_ONBOARDING_FLOW,
        experiment_variation: enabledPreassignExperiment
          ? PREASSIGN_ONBOARDING_EXP_VARIATION
          : CONTROL_EXP_VARIATION,
        action: 'create_content_type',
      });
    }
    tracking.experimentInteracted({
      ...defaultEventProps(),
      experiment_id: FLAGS.EXPERIMENT_PREASSIGN_ONBOARDING_FLOW,
      experiment_variation: enabledPreassignExperiment
        ? PREASSIGN_ONBOARDING_EXP_VARIATION
        : CONTROL_EXP_VARIATION,
      action: 'create_content_type',
    });
  }
};
