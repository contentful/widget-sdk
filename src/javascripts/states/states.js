import accountState from './account';
import spacesState from 'states/Spaces';
import { DeeplinkRouteContainer } from 'states/deeplink/DeeplinkRouteContainer';
import userInvitationState from 'states/UserInvitationState';
import { homeState } from 'features/home';
import { errorState } from './errorState';

import { getModule } from 'core/NgRegistry';

/**
 * @description deeplink route to point users to certain sections,
 * without knowing some details (e.g. navigate to API keys sections,
 * but without specifying their spaceId)
 *
 * Sometimes customer support has to describe "click here and there",
 * in order to explain how to get to some page. This is hard for every-
 * one, both for us and for customers, and this route should solve this
 * exact problem. It infers all parameters from previous usage, or just
 * picks the first one (e.g. first space from all available)
 *
 * @usage
 * https://app.contentful.com/deeplink?link=api
 */

const deeplinkState = {
  name: 'deeplink',
  url: '/deeplink',
  component: DeeplinkRouteContainer,
};

/**
 * Imports all the root states and and adds them to the router.
 * Needs to be called in a 'run' hook to make the application work
 */
export function loadAll() {
  load([
    accountState,
    spacesState,
    homeState,
    deeplinkState,
    userInvitationState,
    errorState,
    {
      name: '_other',
      url: '/*path',
      redirectTo: 'home',
    },
  ]);
}

/**
 * Load only the given states. Used for testing
 */

function load(states) {
  const config = getModule('states/config');
  (states || []).forEach((state) => config.add(state));
  config.init();
}
