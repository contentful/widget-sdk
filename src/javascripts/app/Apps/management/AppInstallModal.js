import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Button, Modal, SelectField, Option } from '@contentful/forma-36-react-components';
import * as Navigator from 'states/Navigator';
import { getOrgSpacesFor, getEnvsFor } from './util';
import { css } from 'emotion';
import tokens from '@contentful/forma-36-tokens';

function goToInstallation(spaceId, environmentId, appId, onClose) {
  Navigator.go({
    path: 'spaces.detail.environment.apps.detail',
    params: {
      spaceId,
      environmentId,
      appId: `private_${appId}`,
      referrer: 'app-management',
    },
    options: {
      location: 'replace',
    },
  }).then(() => onClose());
}

const styles = {
  form: css({
    '> div:first-child': css({
      marginBottom: tokens.spacingL,
    }),
  }),
};

export default function AppInstallModal({ isShown, definition, onClose }) {
  const [redirecting, setRedirecting] = useState(false);
  const [spaces, setSpaces] = useState([]);
  const [spaceEnvs, setSpaceEnvs] = useState([]);
  const [selectedSpace, setSelectedSpace] = useState('');
  const [selectedEnv, setSelectedEnv] = useState('');

  useEffect(() => {
    async function getSpaces() {
      const spaces = await getOrgSpacesFor(definition.sys.organization.sys.id);

      setSpaces(spaces);
      setSelectedSpace(spaces[0].sys.id);
    }

    if (definition) {
      getSpaces();
    }
  }, [definition]);

  useEffect(() => {
    async function getEnvs() {
      const environments = await getEnvsFor(selectedSpace);

      setSpaceEnvs(environments);
      setSelectedEnv(environments[0].sys.id);
    }

    if (selectedSpace !== '') {
      setSelectedEnv('');
      getEnvs();
    }
  }, [selectedSpace]);

  const onContinue = () => {
    setRedirecting(true);
    goToInstallation(selectedSpace, selectedEnv, definition.sys.id, onClose);
  };

  return (
    <Modal
      position="center"
      isShown={isShown}
      title={`Install ${definition.name} to a space`}
      onClose={onClose}>
      {({ title }) => (
        <>
          <Modal.Header title={title} />
          <Modal.Content>
            <div className={styles.form}>
              <SelectField
                labelText="Select a space"
                id="spaceSelection"
                name="spaceSelection"
                required
                value={selectedSpace}
                onChange={(e) => setSelectedSpace(e.target.value)}>
                {spaces.map((space) => (
                  <Option key={space.sys.id} value={space.sys.id}>
                    {space.name}
                  </Option>
                ))}
              </SelectField>
              <SelectField
                labelText="Select an environment"
                id="envSelection"
                name="envSelection"
                required
                onChange={(e) => setSelectedEnv(e.target.value)}
                value={selectedEnv}>
                {spaceEnvs.map((env) => (
                  <Option key={env.sys.id} value={env.sys.id}>
                    {env.name}
                  </Option>
                ))}
              </SelectField>
            </div>
          </Modal.Content>
          <Modal.Controls>
            <Button
              onClick={onContinue}
              buttonType="primary"
              loading={redirecting}
              disabled={redirecting}
              testId="continue-button">
              Continue
            </Button>
            <Button onClick={onClose} buttonType="muted">
              Cancel
            </Button>
          </Modal.Controls>
        </>
      )}
    </Modal>
  );
}

AppInstallModal.propTypes = {
  definition: PropTypes.object,
  onClose: PropTypes.func.isRequired,
  isShown: PropTypes.bool.isRequired,
};
