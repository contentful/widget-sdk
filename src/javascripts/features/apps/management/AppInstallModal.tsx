import React, { useState, useEffect } from 'react';
import { Button, Modal, SelectField, Option } from '@contentful/forma-36-react-components';
import * as Navigator from 'states/Navigator';
import { getOrgSpacesFor, getEnvsFor } from './util';
import { css } from 'emotion';
import tokens from '@contentful/forma-36-tokens';
import { AppDefinition, Space, Environment } from 'contentful-management/types';

function goToInstallation(spaceId, environmentId, appId, onClose) {
  Navigator.go({
    path: 'spaces.environment.apps.detail',
    params: {
      spaceId,
      environmentId,
      appId,
      referrer: 'app-management',
    },
    options: {
      location: 'replace',
    },
  }).then(() => onClose());
}

const styles = {
  form: css({
    '> div:first-child': css({
      marginBottom: tokens.spacingL,
    }),
  }),
};

const getEnvAliases = (env) => {
  if (Array.isArray(env?.sys?.aliases)) {
    return env.sys.aliases;
  }

  return [];
};

const isMasterAliasedToEnv = (env) => {
  return getEnvAliases(env).some((a) => a.sys.id === 'master');
};

const formatEnvName = (env) => {
  const aliases = getEnvAliases(env);

  if (aliases.length) {
    return `${aliases[0].sys.id} > ${env.name}`;
  }

  return env.name;
};

interface AppInstallModalProps {
  definition: AppDefinition;
  onClose: () => void;
  isShown: boolean;
}
export const AppInstallModal: React.FC<AppInstallModalProps> = ({
  isShown,
  definition,
  onClose,
}) => {
  const [redirecting, setRedirecting] = useState(false);
  const [spaces, setSpaces] = useState<Space[]>([]);
  const [spaceEnvs, setSpaceEnvs] = useState<Environment[]>([]);
  const [selectedSpace, setSelectedSpace] = useState('');
  const [selectedEnv, setSelectedEnv] = useState('');

  useEffect(() => {
    async function getSpaces() {
      const spaces = await getOrgSpacesFor(definition.sys.organization.sys.id);

      setSpaces(spaces);
      setSelectedSpace(spaces[0].sys.id);
    }

    if (definition) {
      getSpaces();
    }
  }, [definition]);

  useEffect(() => {
    async function getEnvs() {
      const environments = await getEnvsFor(selectedSpace);

      setSpaceEnvs(environments);
      setSelectedEnv(environments[0].sys.id);
    }

    if (selectedSpace !== '') {
      setSelectedEnv('');
      getEnvs();
    }
  }, [selectedSpace]);

  const onContinue = () => {
    setRedirecting(true);
    const env = spaceEnvs.find((e) => e.sys.id === selectedEnv);

    goToInstallation(
      selectedSpace,
      isMasterAliasedToEnv(env) ? 'master' : selectedEnv,
      definition.sys.id,
      onClose
    );
  };

  return (
    <Modal
      position="center"
      isShown={isShown}
      title={`Install ${definition.name} to a space`}
      onClose={onClose}>
      {({ title }) => (
        <>
          <Modal.Header title={title} />
          <Modal.Content>
            <div className={styles.form}>
              <SelectField
                labelText="Select a space"
                id="spaceSelection"
                name="spaceSelection"
                required
                value={selectedSpace}
                onChange={(e: React.ChangeEvent<HTMLSelectElement>) =>
                  setSelectedSpace(e.target.value)
                }>
                {spaces.map((space) => (
                  <Option key={space.sys.id} value={space.sys.id}>
                    {space.name}
                  </Option>
                ))}
              </SelectField>
              <SelectField
                labelText="Select an environment"
                id="envSelection"
                name="envSelection"
                required
                onChange={(e: React.ChangeEvent<HTMLSelectElement>) =>
                  setSelectedEnv(e.target.value)
                }
                value={selectedEnv}>
                {spaceEnvs.map((env) => (
                  <Option key={env.sys.id} value={env.sys.id}>
                    {formatEnvName(env)}
                  </Option>
                ))}
              </SelectField>
            </div>
          </Modal.Content>
          <Modal.Controls>
            <Button
              onClick={onContinue}
              buttonType="primary"
              loading={redirecting}
              disabled={redirecting}
              testId="continue-button">
              Continue
            </Button>
            <Button onClick={onClose} buttonType="muted">
              Cancel
            </Button>
          </Modal.Controls>
        </>
      )}
    </Modal>
  );
};
