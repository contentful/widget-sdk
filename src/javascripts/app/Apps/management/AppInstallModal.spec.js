import React from 'react';
import { render, wait, cleanup, fireEvent } from '@testing-library/react';
import AppInstallModal from './AppInstallModal';
import * as util from './util';
import * as Navigator from 'states/Navigator';
import mockDefinitions from './mockData/mockDefinitions.json';

jest.mock('./util');

jest.mock('states/Navigator', () => ({
  go: jest.fn()
}));

util.getOrgsAndSpaces = jest.fn(() =>
  Promise.resolve([
    {
      org: { sys: { id: 'my-org-123' } },
      spaces: [
        {
          name: 'mySpace',
          sys: { id: 'my-space-123' },
          organization: { sys: { id: 'my-org-123' } }
        }
      ]
    },
    {
      org: { sys: { id: 'my-org-123' } },
      spaces: [
        {
          name: 'myOtherSpace',
          sys: { id: 'my-other-space-123' },
          organization: { sys: { id: 'my-org-123' } }
        }
      ]
    }
  ])
);

util.getEnvsFor = jest.fn(() =>
  Promise.resolve([
    { name: 'my-env', sys: { id: 'my-env-123' } },
    { name: 'my-other-env', sys: { id: 'my-other-env-123' } }
  ])
);

util.getLastUsedSpace = jest.fn(() => 'my-space-123');

describe('AppInstallModal', () => {
  beforeEach(() => {});

  afterEach(cleanup);

  it('should render the last used space and first env on load and be able to install right away', async () => {
    const wrapper = render(<AppInstallModal definition={mockDefinitions[0]} onClose={() => {}} />);
    await wait();

    expect(wrapper).toMatchSnapshot();

    wrapper.getByTestId('continue-button').click();

    expect(Navigator.go).toHaveBeenCalledWith({
      options: { location: 'replace' },
      params: {
        appId: 'private_3AjEyjWz5tRouW4cVOF9la',
        environmentId: 'my-env-123',
        referrer: 'app-management',
        spaceId: 'my-space-123'
      },
      path: 'spaces.detail.environment.apps.detail'
    });
  });

  it('should be able to pick a different space and env and install', async () => {
    const wrapper = render(<AppInstallModal definition={mockDefinitions[0]} onClose={() => {}} />);
    await wait();

    const [spaceSelector, envSelector] = wrapper.getAllByTestId('cf-ui-select');

    fireEvent.change(spaceSelector, { target: { value: 'my-other-space-123' } });
    await wait();
    fireEvent.change(envSelector, { target: { value: 'my-other-env-123' } });
    await wait();

    wrapper.getByTestId('continue-button').click();

    expect(Navigator.go).toHaveBeenCalledWith({
      options: { location: 'replace' },
      params: {
        appId: 'private_3AjEyjWz5tRouW4cVOF9la',
        environmentId: 'my-other-env-123',
        referrer: 'app-management',
        spaceId: 'my-other-space-123'
      },
      path: 'spaces.detail.environment.apps.detail'
    });
  });
});
