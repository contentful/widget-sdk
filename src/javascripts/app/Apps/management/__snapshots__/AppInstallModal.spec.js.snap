// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInstallModal should render the last used space and first env on load and be able to install right away 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 > div:first-child {
  margin-bottom: 1.5rem;
}

<body
    class="Modal__body--open"
  >
    <div />
    <div
      class="Modal__Modal__portal___F-liP"
    >
      <div
        class="Modal__Modal__overlay___38PZk Modal__Modal__overlay--centered___19mI8 Modal__Modal__overlay--after-open___2WVa7"
      >
        <div
          class="Modal__Modal__wrap___16arr Modal__Modal__wrap--after-open___2mXQz"
          role="dialog"
          style="top: 0px;"
          tabindex="-1"
        >
          <div
            class="Modal__Modal___1064L"
            data-test-id="cf-ui-modal"
            style="width: 520px;"
          >
            <div
              class="ModalHeader__ModalHeader___1yD1S"
              data-test-id="cf-ui-modal-header"
            >
              <h1
                class="ModalHeader__ModalHeader__title___3IuOy"
              >
                Install Google Analytics to a space
              </h1>
            </div>
            <div
              class="ModalContent__ModalContent___2mf3h"
              data-test-id="cf-ui-modal-content"
            >
              <div
                class="emotion-0"
              >
                <div
                  class="SelectField__SelectField___kbQlf"
                  data-test-id="cf-ui-select-field"
                >
                  <div
                    class="SelectField__SelectField__label-wrapper___3jGwo"
                  >
                    <label
                      class="FormLabel__FormLabel___3d6zQ"
                      data-test-id="cf-ui-form-label"
                      for="spaceSelection"
                    >
                      Select a space
                      <span
                        class="FormLabel__FormLabel__required-text___3mvdm"
                      >
                        (
                        required
                        )
                      </span>
                    </label>
                  </div>
                  <div
                    class="Select__Select__wrapper___2pBIz Select__Select--full___7ei7d"
                  >
                    <select
                      aria-label="spaceSelection"
                      class="Select__Select___31Z46 a11y__focus-border--default___60AXp"
                      data-test-id="cf-ui-select"
                      id="spaceSelection"
                      name="spaceSelection"
                      required=""
                    >
                      <option
                        data-test-id="cf-ui-select-option"
                        value="my-space-123"
                      >
                        mySpace
                      </option>
                      <option
                        data-test-id="cf-ui-select-option"
                        value="my-other-space-123"
                      >
                        myOtherSpace
                      </option>
                    </select>
                    <svg
                      class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD Select__Select__icon___boRAZ"
                      data-test-id="cf-ui-icon"
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="SelectField__SelectField___kbQlf"
                  data-test-id="cf-ui-select-field"
                >
                  <div
                    class="SelectField__SelectField__label-wrapper___3jGwo"
                  >
                    <label
                      class="FormLabel__FormLabel___3d6zQ"
                      data-test-id="cf-ui-form-label"
                      for="envSelection"
                    >
                      Select an environment
                      <span
                        class="FormLabel__FormLabel__required-text___3mvdm"
                      >
                        (
                        required
                        )
                      </span>
                    </label>
                  </div>
                  <div
                    class="Select__Select__wrapper___2pBIz Select__Select--full___7ei7d"
                  >
                    <select
                      aria-label="envSelection"
                      class="Select__Select___31Z46 a11y__focus-border--default___60AXp"
                      data-test-id="cf-ui-select"
                      id="envSelection"
                      name="envSelection"
                      required=""
                    >
                      <option
                        data-test-id="cf-ui-select-option"
                        value="my-env-123"
                      >
                        my-env
                      </option>
                      <option
                        data-test-id="cf-ui-select-option"
                        value="my-other-env-123"
                      >
                        my-other-env
                      </option>
                    </select>
                    <svg
                      class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD Select__Select__icon___boRAZ"
                      data-test-id="cf-ui-icon"
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ModalControls__ModalControls___2bTQx"
              data-test-id="cf-ui-modal-controls"
            >
              <button
                class="Button__Button___1ZfFj a11y__focus-border--default___60AXp Button__Button--primary___JImeO"
                data-test-id="continue-button"
                type="button"
              >
                <span
                  class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
                  tabindex="-1"
                >
                  <span
                    class="Button__Button__label___3tcOj"
                  >
                    Continue
                  </span>
                </span>
              </button>
              <button
                class="Button__Button___1ZfFj a11y__focus-border--default___60AXp Button__Button--muted___2Wair"
                data-test-id="cf-ui-button"
                type="button"
              >
                <span
                  class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
                  tabindex="-1"
                >
                  <span
                    class="Button__Button__label___3tcOj"
                  >
                    Cancel
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
